cmd/juju/model/defaultscommand.go

```diff
@@ -97,120 +131,425 @@ func (c *defaultsCommand) SetFlags(f *gnuflag.FlagSet) {
-	if c.api != nil {
-		return c.api, nil
+// handleSetArgs parses args for setting defaults.
+func (c *defaultsCommand) handleSetArgs(args []string) error {
+	argZeroKeyOnly := !strings.Contains(args[0], "=")
+	// If an invalid region was specified, the first positional arg won't have
+	// an "=". If we see one here we know it is invalid.
+	switch {
+	case argZeroKeyOnly && c.regionName == "":
+		return errors.Errorf("invalid region specified: %q", args[0])
+	case argZeroKeyOnly && c.regionName != "":
+		return errors.New("cannot set and retrieve default values simultaneously")
+	default:
+		if err := c.parseSetKeys(args); err != nil {
+			return errors.Trace(err)
+		}
+		c.action = c.setDefaults
+		return nil
 	}
+}
 
-	api, err := c.NewAPIRoot()
-	if err != nil {
-		return nil, errors.Annotate(err, "opening API connection")
+// handleZeroArgs determines whether we are doing a reset only and if any
+// additional arguments are valid -- or why they are not.
+func (c *defaultsCommand) handleZeroArgs() error {
+	resetSpecified := c.resetKeys != nil
+	regionSpecified := c.regionName != ""
+
+	switch {
+	case regionSpecified && !resetSpecified:
+		// It doesn't make sense to specify a region unless we are resetting
+		// values here.
+		return errors.New("specifying a region when retrieving defaults is invalid")
+	case regionSpecified && resetSpecified:
+		// We can reset for a region, c.action stays nil. We want do reset
+		// only.
+		return nil
+	case !regionSpecified && resetSpecified:
+		// Again we want to reset, this is for resetting the controller as
+		// there is no region specified. So c.action remains nil.
+		return nil
+		// case !regionSpecified && !resetSpecified:
+		// This means just get everything, but we should never arrive here if
+		// we short circuited at the begining of c.Init.
+		// c.action = c.getDefaults
+		// return nil
+	default:
+		// This should be unreachable.
+		return errors.New("unexpected input, please file a bug")
 	}
-	client := modelmanager.NewClient(api)
+}
 
-	return client, nil
+// handleOneArg handles the case where we have one positional arg after
+// processing for a region and the reset flag after processing for a region and
+// the reset flagg.
+func (c *defaultsCommand) handleOneArg(arg string) error {
+	resetSpecified := c.resetKeys != nil
+	regionSpecified := c.regionName != ""
+
+	switch {
+	case regionSpecified && !resetSpecified:
+		// It doesn't make sense to specify a region unless we are resetting
+		// values here.
+		return errors.New("specifying a region when retrieving defaults for a setting is invalid")
+	case !regionSpecified && resetSpecified:
+		// It makes no sense to supply a positional arg that isn't a region if
+		// we are resetting a region, so we must have gotten an invalid region.
+		return errors.Errorf("invalid region specified: %q", arg)
+	case !regionSpecified && !resetSpecified:
+		// Phew, we can retrieve a value.
```

Comment:
It would be funny for future readers and editors, too! But I removed the humor:|